pytato/transform/__init__.py

```diff
@@ -259,63 +291,68 @@ def rec(self, expr: ArrayOrNames, *args: P.args, **kwargs: P.kwargs) -> ResultT:
             self._cache[key] = result
             return result
 
+    def rec_function_definition(
+                self, expr: FunctionDefinition, *args: P.args, **kwargs: P.kwargs
+            ) -> FunctionResultT:
+        key = self.get_function_definition_cache_key(expr, *args, **kwargs)
+        try:
+            return self._function_cache[key]
+        except KeyError:
+            result = super().rec_function_definition(expr, *args, **kwargs)
+            self._function_cache[key] = result
+            return result
+
+    def clone_for_callee(
+            self, function: FunctionDefinition) -> Self:
+        """
+        Called to clone *self* before starting traversal of a
+        :class:`pytato.function.FunctionDefinition`.
+        """
+        return type(self)(_function_cache=self._function_cache)
+
 # }}}
 
 
 # {{{ TransformMapper
 
-class TransformMapper(CachedMapper[ArrayOrNames, []]):
+class TransformMapper(CachedMapper[ArrayOrNames, FunctionDefinition, []]):
     """Base class for mappers that transform :class:`pytato.array.Array`\\ s into
     other :class:`pytato.array.Array`\\ s.
 
     Enables certain operations that can only be done if the mapping results are also
     arrays (e.g., calling :meth:`~CachedMapper.get_cache_key` on them). Does not
     implement default mapper methods; for that, see :class:`CopyMapper`.
 
-    .. automethod:: clone_for_callee
     """
+    # FIXME: This inflates recursion depth
     def rec_ary(self, expr: Array) -> Array:
         res = self.rec(expr)
         assert isinstance(res, Array)
         return res
 
-    def clone_for_callee(self, function: FunctionDefinition) -> Self:
-        """
-        Called to clone *self* before starting traversal of a
-        :class:`pytato.function.FunctionDefinition`.
-        """
-        return type(self)()
-
 # }}}
 
 
 # {{{ TransformMapperWithExtraArgs
 
 class TransformMapperWithExtraArgs(
-            CachedMapper[ArrayOrNames, P],
-            Mapper[ArrayOrNames, P]
+            # Why multiple inheriting?
+            CachedMapper[ArrayOrNames, FunctionDefinition, P],
+            Mapper[ArrayOrNames, FunctionDefinition, P]
```

Comment:
```suggestion
            CachedMapper[ArrayOrNames, FunctionDefinition, P],
```
I don't think it should be.

Reply:
Seems to work fine without it.